include/dlaf/matrix/copy.h

```diff
@@ -70,4 +74,79 @@ void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
   copy(source.distribution().localNrTiles(), LocalTileIndex(0, 0), source, LocalTileIndex(0, 0), dest);
 }
 
+/// Copy of a matrix performing data reshuffling
+///
+/// Copy @p src matrix to @p dst, re-distributing data according to the @p dst distribution.
+/// @pre src.size() == dst.size()
+/// @pre equal_process_grid(src, dst)
+/// @pre mb = min(src.blockSize().rows(), dst.blockSize().rows())
+///      src.blockSize().rows() % mb == 0
+///      dst.blockSize().rows() % mb == 0
+/// @pre nb = min(src.blockSize().cols(), dst.blockSize().cols())
+///      src.blockSize().cols() % nb == 0
+///      dst.blockSize().cols() % nb == 0
+template <class T>
+void copy(Matrix<T, Device::CPU>& src,  // TODO this should be const
+          Matrix<T, Device::CPU>& dst, comm::CommunicatorGrid grid) {
+  namespace ex = pika::execution::experimental;
+
+  DLAF_ASSERT_MODERATE(equal_size(src, dst), src.size(), dst.size());
+  DLAF_ASSERT_MODERATE(equal_process_grid(src, grid), src.commGridSize(), grid.size());
+  DLAF_ASSERT_MODERATE(equal_process_grid(dst, grid), dst.commGridSize(), grid.size());
+
+  const TileElementSize tile_size_src = src.blockSize();
+  const TileElementSize tile_size_dst = dst.blockSize();
```

Comment:
I'm not quite sure about the constraints, but should this for now be:
```suggestion
  const TileElementSize tile_size_src = src.baseTileSize();
  const TileElementSize tile_size_dst = dst.baseTileSize();
```
with the possible addition of:
```suggestion
  DLAF_ASSERT(src.blockSize() == src.baseTileSize(), src.blockSize(), src.baseTileSize());
  DLAF_ASSERT(dst.blockSize() == dst.baseTileSize(), dst.blockSize(), dst.baseTileSize());
  const TileElementSize tile_size_src = src.blockSize();
  const TileElementSize tile_size_dst = dst.blockSize();
```

In the future I suppose we could make it:
```suggestion
  const TileElementSize block_size_src = src.blockSize();
  const TileElementSize block_size_dst = dst.blockSize();
```
with the assumption that we can retile a `Matrix` purely based on the block size, independently of what the current tile size is (not relevant until #905, and may not be relevant even after that...).

Reply:
Thanks, makes sense. In that case I would absolutely rename the variable to avoid confusion.

Reply:
Yeah, also makes sense and I see now that I missed some points you were making. Note that I just opened https://github.com/eth-cscs/DLA-Future/pull/935 to handle part of #905. I currently assert that the block and tile sizes are equal, which is implicitly the same requirement that was there for `RetiledMatrix`.

Relaxing the constraints one notch would be to allow retiling already retiled matrices but based on tile size, i.e. only further subdividing the tiles. This is simple to implement, and I'm thinking this may even just work with the current implementation if we remove the assertion.

Relaxing the constraints further would be more complicated. The difficulty with choosing a different retiling is that the retiled matrix doesn't have access to the original matrix anymore which means that you one has to reconstruct tile dependencies. This is simpler if the new retiled tile size is a multiple of the old retiled tile size, and more complicated if it's not. We'll generally have an easier time if we can retile the original matrix instead of retiling a retiled matrix.

That's not to say that we can't or shouldn't do it, but at least from my current knowledge about it the cost seems higher than the value.

Reply:
Sorry for the wall of text. Given the constraint in #935 that a retiled matrix can't be retiled you might not need to check for that here and can instead rely on checks in `RetiledMatrix`/`retiledSubPipeline`.

Further, given the first relaxation would be
> Relaxing the constraints one notch would be to allow retiling already retiled matrices but based on tile size
the right thing to do in this PR is this one:
```suggestion
  const TileElementSize tile_size_src = src.baseTileSize();
  const TileElementSize tile_size_dst = dst.baseTileSize();
```

Reply:
Added also a note https://github.com/eth-cscs/DLA-Future/pull/933/commits/5653f101dedd548c22a3e9068808fca5d16b83c3